{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coding: utf-8\n",
    "from bs4 import BeautifulSoup\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import nltk\n",
    "import re\n",
    "import os\n",
    "import math\n",
    "import json\n",
    "import sys\n",
    "import time\n",
    "import json\n",
    "from PorterStemmer import *\n",
    "#from nltk.stem import PorterStemmer\n",
    "#from nltk.tokenize import word_tokenize\n",
    "ps = PorterStemmer()\n",
    "from tqdm.notebook import tqdm\n",
    "import random\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from nltk.corpus import stopwords\n",
    "stop_words = set(stopwords.words('english'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "delim = '''[ ',(){}.:;\"`\\n]'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"C:\\Files\\col764-a2-release\\col764-ass2-release\\data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = os.path.join(data_dir,\"metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\kshitij alwadhi\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3063: DtypeWarning: Columns (1,4,5,6,13,14,15,16) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "meta_df = pd.read_csv(metadata)\n",
    "meta_df = meta_df[['cord_uid','title','abstract','pdf_json_files','pmc_json_files']]\n",
    "#meta_df = meta_df.head(1000)\n",
    "meta_df['pdf_json_files'] = meta_df['pdf_json_files'].astype(str)\n",
    "meta_df['pmc_json_files'] = meta_df['pmc_json_files'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>pdf_json_files</th>\n",
       "      <th>pmc_json_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ug7v899j</td>\n",
       "      <td>Clinical features of culture-proven Mycoplasma...</td>\n",
       "      <td>OBJECTIVE: This retrospective chart review des...</td>\n",
       "      <td>document_parses/pdf_json/d1aafb70c066a2068b027...</td>\n",
       "      <td>document_parses/pmc_json/PMC35282.xml.json</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>02tnwd4m</td>\n",
       "      <td>Nitric oxide: a pro-inflammatory mediator in l...</td>\n",
       "      <td>Inflammatory diseases of the respiratory tract...</td>\n",
       "      <td>document_parses/pdf_json/6b0567729c2143a66d737...</td>\n",
       "      <td>document_parses/pmc_json/PMC59543.xml.json</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ejv2xln0</td>\n",
       "      <td>Surfactant protein-D and pulmonary host defense</td>\n",
       "      <td>Surfactant protein-D (SP-D) participates in th...</td>\n",
       "      <td>document_parses/pdf_json/06ced00a5fc04215949aa...</td>\n",
       "      <td>document_parses/pmc_json/PMC59549.xml.json</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2b73a28n</td>\n",
       "      <td>Role of endothelin-1 in lung disease</td>\n",
       "      <td>Endothelin-1 (ET-1) is a 21 amino acid peptide...</td>\n",
       "      <td>document_parses/pdf_json/348055649b6b8cf2b9a37...</td>\n",
       "      <td>document_parses/pmc_json/PMC59574.xml.json</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9785vg6d</td>\n",
       "      <td>Gene expression in epithelial cells in respons...</td>\n",
       "      <td>Respiratory syncytial virus (RSV) and pneumoni...</td>\n",
       "      <td>document_parses/pdf_json/5f48792a5fa08bed9f560...</td>\n",
       "      <td>document_parses/pmc_json/PMC59580.xml.json</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid                                              title  \\\n",
       "0  ug7v899j  Clinical features of culture-proven Mycoplasma...   \n",
       "1  02tnwd4m  Nitric oxide: a pro-inflammatory mediator in l...   \n",
       "2  ejv2xln0    Surfactant protein-D and pulmonary host defense   \n",
       "3  2b73a28n               Role of endothelin-1 in lung disease   \n",
       "4  9785vg6d  Gene expression in epithelial cells in respons...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  OBJECTIVE: This retrospective chart review des...   \n",
       "1  Inflammatory diseases of the respiratory tract...   \n",
       "2  Surfactant protein-D (SP-D) participates in th...   \n",
       "3  Endothelin-1 (ET-1) is a 21 amino acid peptide...   \n",
       "4  Respiratory syncytial virus (RSV) and pneumoni...   \n",
       "\n",
       "                                      pdf_json_files  \\\n",
       "0  document_parses/pdf_json/d1aafb70c066a2068b027...   \n",
       "1  document_parses/pdf_json/6b0567729c2143a66d737...   \n",
       "2  document_parses/pdf_json/06ced00a5fc04215949aa...   \n",
       "3  document_parses/pdf_json/348055649b6b8cf2b9a37...   \n",
       "4  document_parses/pdf_json/5f48792a5fa08bed9f560...   \n",
       "\n",
       "                               pmc_json_files  \n",
       "0  document_parses/pmc_json/PMC35282.xml.json  \n",
       "1  document_parses/pmc_json/PMC59543.xml.json  \n",
       "2  document_parses/pmc_json/PMC59549.xml.json  \n",
       "3  document_parses/pmc_json/PMC59574.xml.json  \n",
       "4  document_parses/pmc_json/PMC59580.xml.json  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getTokensFromText(text):\n",
    "    words = re.split(delim, text)\n",
    "    res = []\n",
    "    for w in words:\n",
    "        if(len(w) > 0):\n",
    "            #temp = ps.stem(w.lower(), 0, len(w)-1)\n",
    "            temp = w.lower()\n",
    "            if temp not in stop_words:\n",
    "                res.append(temp)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_idf_dict():\n",
    "    idf_dict = {}\n",
    "    filemap = {}\n",
    "    for i,row in tqdm(meta_df.iterrows()):\n",
    "        uid = row['cord_uid']\n",
    "        alltext = \"\"\n",
    "        if row['pmc_json_files']:\n",
    "            paths = row['pmc_json_files']\n",
    "        elif row['pdf_json_files']:\n",
    "            paths = row['pdf_json_files']\n",
    "        filemap[uid] = paths\n",
    "        for path in paths.split(\"; \"):\n",
    "            if(path == \"nan\"):\n",
    "                continue\n",
    "            datapath = os.path.join(data_dir,path)\n",
    "            with open(datapath) as data:\n",
    "                data_dict = json.load(data)\n",
    "                for section in data_dict['body_text']:\n",
    "                    section_text = section['text']\n",
    "                    alltext+=section_text\n",
    "                    alltext+=\" \"\n",
    "        tokens = getTokensFromText(alltext)\n",
    "        tokens = set(tokens)\n",
    "        for token in tokens:\n",
    "            if token not in idf_dict:\n",
    "                idf_dict[token]=1\n",
    "            else:\n",
    "                idf_dict[token]+=1\n",
    "    return idf_dict,filemap        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_idf_dict_new():\n",
    "    idf_dict = {}\n",
    "    filemap = {}\n",
    "    for row in tqdm(meta_df.itertuples(),total = meta_df.shape[0]):\n",
    "        uid = row.cord_uid\n",
    "        alltext = \"\"\n",
    "        if row.pmc_json_files:\n",
    "            paths = row.pmc_json_files\n",
    "        elif row.pdf_json_files:\n",
    "            paths = row.pdf_json_files\n",
    "        filemap[uid] = paths\n",
    "        for path in paths.split(\"; \"):\n",
    "            if(path == \"nan\"):\n",
    "                continue\n",
    "            datapath = os.path.join(data_dir,path)\n",
    "            with open(datapath) as data:\n",
    "                data_dict = json.load(data)\n",
    "                for section in data_dict['body_text']:\n",
    "                    section_text = section['text']\n",
    "                    alltext+=section_text\n",
    "                    alltext+=\" \"\n",
    "        tokens = getTokensFromText(alltext)\n",
    "        tokens = set(tokens)\n",
    "        for token in tokens:\n",
    "            if token not in idf_dict:\n",
    "                idf_dict[token]=1\n",
    "            else:\n",
    "                idf_dict[token]+=1\n",
    "    return idf_dict,filemap        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "93252b86e1ea4ee5a20f1cbf01e484ca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=1000.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "idf_dict,filemap = get_idf_dict_new()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fbc5239c5a75496db25acbcb4ddcff24",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=192509.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "def get_filepath():\n",
    "    filemap = {}\n",
    "    meta_df2 = pd.read_csv(metadata)\n",
    "    meta_df2 = meta_df2[['cord_uid','title','abstract','pdf_json_files','pmc_json_files']]\n",
    "    meta_df2['pdf_json_files'] = meta_df2['pdf_json_files'].astype(str)\n",
    "    meta_df2['pmc_json_files'] = meta_df2['pmc_json_files'].astype(str)\n",
    "    for row in tqdm(meta_df2.itertuples(),total = meta_df2.shape[0]):\n",
    "        uid = row.cord_uid\n",
    "        alltext = \"\"\n",
    "        if row.pmc_json_files:\n",
    "            paths = row.pmc_json_files\n",
    "        elif row.pdf_json_files:\n",
    "            paths = row.pdf_json_files\n",
    "        filemap[uid] = paths\n",
    "    return filemap        \n",
    "filemap = get_filepath()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dumpdicts(idf_dict,filemap):\n",
    "    with open(\"idf_dict.json\", \"w\") as outfile:\n",
    "        json.dump(idf_dict, outfile)\n",
    "    with open(\"filemap.json\", \"w\") as outfile:\n",
    "        json.dump(filemap, outfile)\n",
    "dumpdicts(idf_dict,filemap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"idf_dict.json\", \"r\") as jsonfile:\n",
    "    idf_dict = json.load(jsonfile)\n",
    "with open(\"filemap.json\", \"r\") as jsonfile:\n",
    "    filemap = json.load(jsonfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_vocab_dict(idf_dict):\n",
    "    vocab = {}\n",
    "    for i,word in enumerate(idf_dict.keys()):\n",
    "        vocab[word] = i\n",
    "    return vocab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab = create_vocab_dict(idf_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'exudation': 0,\n",
       " 'admission': 1,\n",
       " 'winter': 2,\n",
       " 'symptoms': 3,\n",
       " 'hoped': 4,\n",
       " 'spinal': 5,\n",
       " 'rarely': 6,\n",
       " 'categorical': 7,\n",
       " 'frequently': 8,\n",
       " 'radiological': 9,\n",
       " 'include': 10,\n",
       " '27': 11,\n",
       " 'pleurisy': 12,\n",
       " 'approach': 13,\n",
       " '75%': 14,\n",
       " 'confirmed': 15,\n",
       " 'weeks': 16,\n",
       " 'malignancy': 17,\n",
       " 'except': 18,\n",
       " '7]': 19,\n",
       " 'another': 20,\n",
       " 'king': 21,\n",
       " 'p': 22,\n",
       " 'kpa': 23,\n",
       " 'severe': 24,\n",
       " 'mortality': 25,\n",
       " 'thirty-three': 26,\n",
       " 'viruses': 27,\n",
       " 'usually': 28,\n",
       " 'physical': 29,\n",
       " 'cell': 30,\n",
       " '30': 31,\n",
       " 'nerve': 32,\n",
       " 'uneventful': 33,\n",
       " 'diagnosis': 34,\n",
       " 'present': 35,\n",
       " 'tests': 36,\n",
       " 'organism': 37,\n",
       " 'possible': 38,\n",
       " 'populations': 39,\n",
       " 'schools': 40,\n",
       " 'meningoencephalitis': 41,\n",
       " 'hemorrhages': 42,\n",
       " 'bias': 43,\n",
       " '87': 44,\n",
       " 'palsy': 45,\n",
       " 'methods': 46,\n",
       " 'bronchial': 47,\n",
       " 'microbiology': 48,\n",
       " 'although': 49,\n",
       " 'caused': 50,\n",
       " '35%': 51,\n",
       " 'hours': 52,\n",
       " 'prompted': 53,\n",
       " 'manifestations': 54,\n",
       " 'france': 55,\n",
       " 'bronchiolitis': 56,\n",
       " 'lung': 57,\n",
       " 'viral': 58,\n",
       " 'sign': 59,\n",
       " 'required': 60,\n",
       " 'pleural': 61,\n",
       " '3%': 62,\n",
       " '77': 63,\n",
       " 'al-qassim': 64,\n",
       " 'rare': 65,\n",
       " 'psychosis': 66,\n",
       " 'obtained': 67,\n",
       " 'identified': 68,\n",
       " 'arterial': 69,\n",
       " 'disorder': 70,\n",
       " '16/31': 71,\n",
       " '11': 72,\n",
       " 'radiographic': 73,\n",
       " 'among': 74,\n",
       " '-10%': 75,\n",
       " 'fermentation': 76,\n",
       " 'virtually': 77,\n",
       " 'involve': 78,\n",
       " 'organisms': 79,\n",
       " 'based': 80,\n",
       " 'diameter': 81,\n",
       " '∝': 82,\n",
       " 'adults': 83,\n",
       " 'morphology': 84,\n",
       " '21/24': 85,\n",
       " 'military': 86,\n",
       " 'kauh': 87,\n",
       " 'categories': 88,\n",
       " 'majority': 89,\n",
       " '32': 90,\n",
       " 'presenting': 91,\n",
       " 'seen': 92,\n",
       " 'statistic': 93,\n",
       " 'capacity': 94,\n",
       " 'wheezes': 95,\n",
       " 'asymptomatic': 96,\n",
       " 'coagulation': 97,\n",
       " 'hospital': 98,\n",
       " 'congenial': 99,\n",
       " 'tertiary': 100,\n",
       " 'classic': 101,\n",
       " 'distress': 102,\n",
       " 'died': 103,\n",
       " 'standard': 104,\n",
       " 'high-grade': 105,\n",
       " 'second': 106,\n",
       " 'pneumoniae': 107,\n",
       " 'collection': 108,\n",
       " '<90': 109,\n",
       " 'obstructive': 110,\n",
       " '1997': 111,\n",
       " 'groups': 112,\n",
       " 'endotracheal': 113,\n",
       " 'eleven': 114,\n",
       " 'adult': 115,\n",
       " 'occurred': 116,\n",
       " 'effusion': 117,\n",
       " '1': 118,\n",
       " 'defined': 119,\n",
       " 'spring': 120,\n",
       " '265': 121,\n",
       " '9': 122,\n",
       " 'tachypnoea': 123,\n",
       " 'outpatient': 124,\n",
       " 'person-to-person': 125,\n",
       " '6': 126,\n",
       " 'however': 127,\n",
       " 'myocarditis': 128,\n",
       " 'disease': 129,\n",
       " '4/4': 130,\n",
       " 'isolated': 131,\n",
       " 'fifth': 132,\n",
       " 'transmissible': 133,\n",
       " 'nosocomial': 134,\n",
       " '[8]': 135,\n",
       " 'cardiovascular': 136,\n",
       " 'sick': 137,\n",
       " 'related': 138,\n",
       " '2%': 139,\n",
       " 'pharyngitis': 140,\n",
       " 'close': 141,\n",
       " 'respectively': 142,\n",
       " 'explanation': 143,\n",
       " 'three': 144,\n",
       " 'blood': 145,\n",
       " 'pleuritic': 146,\n",
       " 'typical': 147,\n",
       " '92': 148,\n",
       " 'january': 149,\n",
       " 'antibiotic': 150,\n",
       " 'finding': 151,\n",
       " 'camps': 152,\n",
       " 'white': 153,\n",
       " 'erythema': 154,\n",
       " 'closed': 155,\n",
       " 'least': 156,\n",
       " 'intravascular': 157,\n",
       " 'classified': 158,\n",
       " '<90%': 159,\n",
       " 'teaching': 160,\n",
       " 'community-acquired': 161,\n",
       " '6%': 162,\n",
       " 'cough': 163,\n",
       " 'acute': 164,\n",
       " 'premature': 165,\n",
       " 'causes': 166,\n",
       " 'turbidity': 167,\n",
       " '20': 168,\n",
       " 'contact': 169,\n",
       " '5%': 170,\n",
       " 'occurs': 171,\n",
       " 'hematologic': 172,\n",
       " 'infant': 173,\n",
       " '18000': 174,\n",
       " 'colonies': 175,\n",
       " '10-150': 176,\n",
       " 'clinicians': 177,\n",
       " '567': 178,\n",
       " 'arabia': 179,\n",
       " 'protean': 180,\n",
       " 'data': 181,\n",
       " 'culture': 182,\n",
       " '57': 183,\n",
       " 'annual': 184,\n",
       " 'heart': 185,\n",
       " 'atrophy': 186,\n",
       " 'sarcoidosis': 187,\n",
       " 'emergency': 188,\n",
       " 'recruits': 189,\n",
       " 'sounds': 190,\n",
       " 'underlying': 191,\n",
       " 'breathing': 192,\n",
       " 'infants': 193,\n",
       " 'infected': 194,\n",
       " 'included': 195,\n",
       " 'g': 196,\n",
       " 'values': 197,\n",
       " 'cause': 198,\n",
       " 'orange': 199,\n",
       " 'simplex': 200,\n",
       " 'cloacae': 201,\n",
       " '72': 202,\n",
       " 'disc': 203,\n",
       " 'comorbidities': 204,\n",
       " 'results': 205,\n",
       " 'igm': 206,\n",
       " 'culture-proven': 207,\n",
       " 'tract': 208,\n",
       " 'immunodeficiency': 209,\n",
       " 'steroid': 210,\n",
       " 'atypical': 211,\n",
       " '30%': 212,\n",
       " 'since': 213,\n",
       " '1998': 214,\n",
       " 'subjects': 215,\n",
       " 'grow': 216,\n",
       " 'published': 217,\n",
       " 'considered': 218,\n",
       " 'similarly': 219,\n",
       " 'nervous': 220,\n",
       " 'syndrome': 221,\n",
       " 'singly': 222,\n",
       " 'shock': 223,\n",
       " 'serological': 224,\n",
       " 'color': 225,\n",
       " 'third': 226,\n",
       " 'episodes': 227,\n",
       " 'optic': 228,\n",
       " 'laboratory': 229,\n",
       " 'therefore': 230,\n",
       " 'features': 231,\n",
       " 'pneumothorax': 232,\n",
       " 'care': 233,\n",
       " 'school': 234,\n",
       " 'long': 235,\n",
       " 'consolidation': 236,\n",
       " 'lasting': 237,\n",
       " 'droplets': 238,\n",
       " 'moderate': 239,\n",
       " 'remaining': 240,\n",
       " 'area': 241,\n",
       " 'prematurely': 242,\n",
       " 'aslo': 243,\n",
       " 'transmitted': 244,\n",
       " 'collected': 245,\n",
       " 'package': 246,\n",
       " 'tendency': 247,\n",
       " 'cytomegalo': 248,\n",
       " 'asthma': 249,\n",
       " 'treated': 250,\n",
       " '19': 251,\n",
       " 'microbio': 252,\n",
       " 'mmhg': 253,\n",
       " '7%': 254,\n",
       " 'sciences': 255,\n",
       " 'profile': 256,\n",
       " '=': 257,\n",
       " 'edema': 258,\n",
       " '[11]': 259,\n",
       " 'paper': 260,\n",
       " 'occur': 261,\n",
       " 'infections': 262,\n",
       " 'failure': 263,\n",
       " 'acquired': 264,\n",
       " 'tracheobronchitis': 265,\n",
       " 'medium': 266,\n",
       " 'larger': 267,\n",
       " '[9]': 268,\n",
       " 'demyelinating': 269,\n",
       " 'high': 270,\n",
       " '12': 271,\n",
       " 'tachycardia': 272,\n",
       " 'cultures': 273,\n",
       " 'aged': 274,\n",
       " 'frequent': 275,\n",
       " 'slightly': 276,\n",
       " 'records': 277,\n",
       " 'small': 278,\n",
       " '[6': 279,\n",
       " 'table': 280,\n",
       " 'clinically': 281,\n",
       " '5]': 282,\n",
       " 'fisher': 283,\n",
       " 'eye': 284,\n",
       " 'radiologically': 285,\n",
       " 'college': 286,\n",
       " 'aspiration': 287,\n",
       " 'recovered': 288,\n",
       " 'myalgias': 289,\n",
       " 'pain': 290,\n",
       " 'require': 291,\n",
       " 'colonial': 292,\n",
       " 'infection': 293,\n",
       " 'death': 294,\n",
       " 'fastidious': 295,\n",
       " 'disparity': 296,\n",
       " 'temperate': 297,\n",
       " 'acinetobacter': 298,\n",
       " 'recovery': 299,\n",
       " 'hypotension': 300,\n",
       " 'available': 301,\n",
       " 'fibrosis': 302,\n",
       " 'opposed': 303,\n",
       " 'agent': 304,\n",
       " '>30/minute': 305,\n",
       " 'febrile': 306,\n",
       " 'neuropathies': 307,\n",
       " 'centers': 308,\n",
       " '5-23%': 309,\n",
       " 'hemolytic': 310,\n",
       " 'former': 311,\n",
       " 'sickle': 312,\n",
       " 'epidemiological': 313,\n",
       " '[1': 314,\n",
       " 'muscle': 315,\n",
       " 'social': 316,\n",
       " 'clinical': 317,\n",
       " '82': 318,\n",
       " 'guillain-barre': 319,\n",
       " 'pneumofast®': 320,\n",
       " 'cultured': 321,\n",
       " 'particularly': 322,\n",
       " 'characteristics': 323,\n",
       " '25%': 324,\n",
       " 'diagnosed': 325,\n",
       " 'epidemiology': 326,\n",
       " '14': 327,\n",
       " 'severity': 328,\n",
       " 'newborns': 329,\n",
       " 'pre-publication': 330,\n",
       " 'pathogens': 331,\n",
       " '<8': 332,\n",
       " 'secretion': 333,\n",
       " 'inpatients': 334,\n",
       " 'unhospitalized': 335,\n",
       " 'countries': 336,\n",
       " 'retinal': 337,\n",
       " 'enough': 338,\n",
       " 'syncytial': 339,\n",
       " 'concerning': 340,\n",
       " '[8': 341,\n",
       " '7': 342,\n",
       " 'arthralgias': 343,\n",
       " 'system': 344,\n",
       " 'rate': 345,\n",
       " '17': 346,\n",
       " '21': 347,\n",
       " 'examination': 348,\n",
       " '9%': 349,\n",
       " 'processed': 350,\n",
       " 'specimens': 351,\n",
       " 'systolic': 352,\n",
       " 'following': 353,\n",
       " 'congenital': 354,\n",
       " 'mild': 355,\n",
       " 'human': 356,\n",
       " 'evan': 357,\n",
       " 'difficult': 358,\n",
       " 'associated': 359,\n",
       " 'presented': 360,\n",
       " 'community': 361,\n",
       " 'mycoplasma': 362,\n",
       " 'known': 363,\n",
       " '[10]': 364,\n",
       " '15': 365,\n",
       " '8': 366,\n",
       " '18]': 367,\n",
       " 'facility': 368,\n",
       " 'virus': 369,\n",
       " 'identification': 370,\n",
       " 'bronchopneumonia': 371,\n",
       " 'fall': 372,\n",
       " 'areas': 373,\n",
       " 'cases': 374,\n",
       " '>140': 375,\n",
       " 'thirty-one': 376,\n",
       " 'group': 377,\n",
       " 'crepitations': 378,\n",
       " 'causative': 379,\n",
       " '3-months': 380,\n",
       " 'concomitantly': 381,\n",
       " 'exact': 382,\n",
       " '83': 383,\n",
       " 'therapy': 384,\n",
       " 'pneumonia': 385,\n",
       " '[6]': 386,\n",
       " 'spss': 387,\n",
       " 'comparable': 388,\n",
       " 'positive': 389,\n",
       " 'shown': 390,\n",
       " 'pneumonic': 391,\n",
       " 'found': 392,\n",
       " 'riyadh': 393,\n",
       " 'standardized': 394,\n",
       " 'admissions': 395,\n",
       " 'dermatologic': 396,\n",
       " 'thrombocytopenia': 397,\n",
       " '79': 398,\n",
       " 'beyond': 399,\n",
       " '2': 400,\n",
       " 'students': 401,\n",
       " 'december': 402,\n",
       " 'urticarial': 403,\n",
       " 'none': 404,\n",
       " 'musculoskeletal': 405,\n",
       " 'fourth': 406,\n",
       " 'relatively': 407,\n",
       " 'demographic': 408,\n",
       " 'climates': 409,\n",
       " 'obacter': 410,\n",
       " 'techniques': 411,\n",
       " 'fever': 412,\n",
       " 'likely': 413,\n",
       " 'old': 414,\n",
       " 'used': 415,\n",
       " 'pericarditis': 416,\n",
       " 'arthritis': 417,\n",
       " 'septic': 418,\n",
       " 'instructions': 419,\n",
       " 'anaerobically': 420,\n",
       " 'age': 421,\n",
       " 'scarce': 422,\n",
       " 'dry': 423,\n",
       " 'rashes': 424,\n",
       " '112': 425,\n",
       " 'wished': 426,\n",
       " 'occasion': 427,\n",
       " 'perform': 428,\n",
       " 'abdulaziz': 429,\n",
       " 'affected': 430,\n",
       " '3': 431,\n",
       " 'productive': 432,\n",
       " 'bed': 433,\n",
       " 'glucose': 434,\n",
       " 'lower': 435,\n",
       " 'multiforme': 436,\n",
       " 'patients': 437,\n",
       " 'myeloproliferative': 438,\n",
       " '10%': 439,\n",
       " 'remains': 440,\n",
       " 'immunocompromised': 441,\n",
       " 'versus': 442,\n",
       " 'agars': 443,\n",
       " 'broth': 444,\n",
       " 'less': 445,\n",
       " 'signs': 446,\n",
       " '37': 447,\n",
       " 'admitted': 448,\n",
       " '4': 449,\n",
       " 'occasions': 450,\n",
       " 'beds': 451,\n",
       " 'cardiopulmonary': 452,\n",
       " 'agar': 453,\n",
       " 'enter': 454,\n",
       " 'malaise': 455,\n",
       " 'management': 456,\n",
       " '60': 457,\n",
       " 'inspected': 458,\n",
       " 'expected': 459,\n",
       " '[12]': 460,\n",
       " '24-months': 461,\n",
       " 'year-round': 462,\n",
       " 'lavage': 463,\n",
       " '60%': 464,\n",
       " '39°c': 465,\n",
       " 'amplification': 466,\n",
       " 'jeddah': 467,\n",
       " 'species': 468,\n",
       " '16': 469,\n",
       " 'pressure': 470,\n",
       " 'young': 471,\n",
       " 'according': 472,\n",
       " 'using': 473,\n",
       " 'accounted': 474,\n",
       " 'program': 475,\n",
       " '70': 476,\n",
       " 'may': 477,\n",
       " 'chest': 478,\n",
       " 'change': 479,\n",
       " 'gestation': 480,\n",
       " 'statistical': 481,\n",
       " 'outcome': 482,\n",
       " 'information': 483,\n",
       " 'additionally': 484,\n",
       " 'hypoxemia': 485,\n",
       " '20/24': 486,\n",
       " 'fried-egg-like': 487,\n",
       " 'unrelated': 488,\n",
       " '%': 489,\n",
       " '19000': 490,\n",
       " 'igg': 491,\n",
       " 'bronchoalveolar': 492,\n",
       " 'organ': 493,\n",
       " 'common': 494,\n",
       " 'also': 495,\n",
       " 'chain': 496,\n",
       " 'outpatients': 497,\n",
       " 'presentation': 498,\n",
       " 'croup': 499,\n",
       " 'saudi': 500,\n",
       " 'detected': 501,\n",
       " 'and/or': 502,\n",
       " 'light': 503,\n",
       " 'difference': 504,\n",
       " 'complications': 505,\n",
       " 'shed': 506,\n",
       " 'accessed': 507,\n",
       " 'sputum': 508,\n",
       " 'chi-square': 509,\n",
       " 'rsv': 510,\n",
       " 'outbreaks': 511,\n",
       " 'signes': 512,\n",
       " 'history': 513,\n",
       " 'summer': 514,\n",
       " 'partial': 515,\n",
       " 'twenty-four': 516,\n",
       " 'develop': 517,\n",
       " '40': 518,\n",
       " '8/9': 519,\n",
       " 'e': 520,\n",
       " '56': 521,\n",
       " 'bacterial': 522,\n",
       " 'contrasts': 523,\n",
       " 'conclusion': 524,\n",
       " 'congestive': 525,\n",
       " 'along': 526,\n",
       " 'boarding': 527,\n",
       " 'international': 528,\n",
       " 'bacteria': 529,\n",
       " 'incubated': 530,\n",
       " 'previously': 531,\n",
       " 'two': 532,\n",
       " 'consistent': 533,\n",
       " 'perhaps': 534,\n",
       " '05': 535,\n",
       " 'performed': 536,\n",
       " '22': 537,\n",
       " 'comparison': 538,\n",
       " 'immunity': 539,\n",
       " 'cerebral': 540,\n",
       " '40%': 541,\n",
       " 'total': 542,\n",
       " 'due': 543,\n",
       " 'born': 544,\n",
       " '5': 545,\n",
       " 'maculopapular': 546,\n",
       " 'oxygen': 547,\n",
       " 'summarized': 548,\n",
       " 'non-immunocompromised': 549,\n",
       " 'trays': 550,\n",
       " 'minor': 551,\n",
       " 'exposure': 552,\n",
       " '37°c': 553,\n",
       " 'incubation': 554,\n",
       " 'period': 555,\n",
       " '20/36': 556,\n",
       " 'within': 557,\n",
       " 'polymerase': 558,\n",
       " '511': 559,\n",
       " '/minute': 560,\n",
       " 'granular': 561,\n",
       " 'older': 562,\n",
       " 'gram-stained': 563,\n",
       " 'study': 564,\n",
       " 'university': 565,\n",
       " 'years': 566,\n",
       " '0': 567,\n",
       " '[13]': 568,\n",
       " '8%': 569,\n",
       " 'selection': 570,\n",
       " 'departments': 571,\n",
       " 'sensitivity': 572,\n",
       " 'preschool': 573,\n",
       " 'nodosum': 574,\n",
       " 'retrospective': 575,\n",
       " 'test': 576,\n",
       " 'highly': 577,\n",
       " 'nasopharyngeal': 578,\n",
       " 'elderly': 579,\n",
       " 'half': 580,\n",
       " 'myelodysplastic': 581,\n",
       " '23%': 582,\n",
       " 'charts': 583,\n",
       " 'evidence': 584,\n",
       " 'abscess': 585,\n",
       " 'yellow': 586,\n",
       " 'tray': 587,\n",
       " 'adenovirus': 588,\n",
       " '24': 589,\n",
       " 'breath': 590,\n",
       " '10]': 591,\n",
       " 'one': 592,\n",
       " 'family': 593,\n",
       " 'airway': 594,\n",
       " 'type': 595,\n",
       " 'average': 596,\n",
       " 'school-aged': 597,\n",
       " 'review': 598,\n",
       " 'daily': 599,\n",
       " 'analysis': 600,\n",
       " 'assist': 601,\n",
       " 'herpes': 602,\n",
       " 'absence': 603,\n",
       " 'respiratory': 604,\n",
       " '58': 605,\n",
       " 'pneumofast': 606,\n",
       " 'children': 607,\n",
       " 'reaction-gene': 608,\n",
       " 'isolates': 609,\n",
       " 'recruit': 610,\n",
       " 'upper': 611,\n",
       " 'manufacturer': 612,\n",
       " 'attributable': 613,\n",
       " 'anemia': 614,\n",
       " 'reviewed': 615,\n",
       " 'demographics': 616,\n",
       " 'accounting': 617,\n",
       " 'saturation': 618,\n",
       " 'red': 619,\n",
       " 'adhesion': 620,\n",
       " '[27': 621,\n",
       " 'epithelial': 622,\n",
       " 'activation': 623,\n",
       " 'materials': 624,\n",
       " 'affects': 625,\n",
       " 'basis': 626,\n",
       " 'notion': 627,\n",
       " 'heme': 628,\n",
       " 'dna': 629,\n",
       " 'ample': 630,\n",
       " 'phagocyte-generated': 631,\n",
       " 'active': 632,\n",
       " 'pc20': 633,\n",
       " 'approaches': 634,\n",
       " 'together': 635,\n",
       " 's-nitrosylation': 636,\n",
       " 'guanylyl': 637,\n",
       " 'sulfation': 638,\n",
       " 'nos': 639,\n",
       " 'peroxynitrite': 640,\n",
       " '28': 641,\n",
       " 'deletion': 642,\n",
       " 'hyper-reactivity': 643,\n",
       " 'result': 644,\n",
       " 'proof': 645,\n",
       " '[46]': 646,\n",
       " 'postulated': 647,\n",
       " 'minimized': 648,\n",
       " 'factor-κb': 649,\n",
       " 'altered': 650,\n",
       " '34]': 651,\n",
       " 'per': 652,\n",
       " 'interpreting': 653,\n",
       " 'levels': 654,\n",
       " 'assessment': 655,\n",
       " 'regulation': 656,\n",
       " 'allergic': 657,\n",
       " 'crucial': 658,\n",
       " 'still': 659,\n",
       " '106': 660,\n",
       " 'minimize': 661,\n",
       " 'caution': 662,\n",
       " 'merely': 663,\n",
       " 'described': 664,\n",
       " 'superoxide': 665,\n",
       " 'cytokines': 666,\n",
       " '38': 667,\n",
       " 'extensive': 668,\n",
       " 'contrasting': 669,\n",
       " 'humans': 670,\n",
       " 'increasing': 671,\n",
       " 'protein-associated': 672,\n",
       " 'conclusive': 673,\n",
       " 'samples': 674,\n",
       " 'indirect': 675,\n",
       " 'discovery': 676,\n",
       " 'estimates': 677,\n",
       " 'without': 678,\n",
       " 'oxidization': 679,\n",
       " '[13': 680,\n",
       " 'endothelial': 681,\n",
       " 'cysteine': 682,\n",
       " 'disruption': 683,\n",
       " 'intriguingly': 684,\n",
       " 'proteins': 685,\n",
       " 'tumor': 686,\n",
       " '[17]': 687,\n",
       " 'scavengers': 688,\n",
       " 'induce': 689,\n",
       " 'cystic': 690,\n",
       " 'increased': 691,\n",
       " 'epo-deficient': 692,\n",
       " 'mutation': 693,\n",
       " 'extent': 694,\n",
       " 'significantly': 695,\n",
       " 'vliet': 696,\n",
       " 'monitoring': 697,\n",
       " 'site': 698,\n",
       " 'latter': 699,\n",
       " 'notably': 700,\n",
       " 'recognized': 701,\n",
       " 'inoculation': 702,\n",
       " 'largely': 703,\n",
       " 'given': 704,\n",
       " 'relation': 705,\n",
       " 'indicating': 706,\n",
       " 'affecting': 707,\n",
       " 'relative': 708,\n",
       " 'myeloperoxidase': 709,\n",
       " 'much': 710,\n",
       " 'possibly': 711,\n",
       " '29': 712,\n",
       " 'implicate': 713,\n",
       " 'endotoxin-induced': 714,\n",
       " 'activated': 715,\n",
       " 'preliminary': 716,\n",
       " 'diffusion-limited': 717,\n",
       " 'investigators': 718,\n",
       " 'onoo-[16]': 719,\n",
       " 'accounts': 720,\n",
       " '30]': 721,\n",
       " 'alveolar': 722,\n",
       " 'markers': 723,\n",
       " 'relate': 724,\n",
       " 'polymorphism': 725,\n",
       " 'sources': 726,\n",
       " 'possibility': 727,\n",
       " 'decreases': 728,\n",
       " '[40]': 729,\n",
       " '1-1000': 730,\n",
       " 'would': 731,\n",
       " 'cyclase': 732,\n",
       " 'full': 733,\n",
       " 'target': 734,\n",
       " 'pathology': 735,\n",
       " 'brings': 736,\n",
       " 'regarded': 737,\n",
       " 'pro-injurious': 738,\n",
       " '31': 739,\n",
       " 'ischemia/reperfusion': 740,\n",
       " 'elevated': 741,\n",
       " 'inflammatory': 742,\n",
       " 'macrophages': 743,\n",
       " 'linkage': 744,\n",
       " 'quantitative': 745,\n",
       " 'cytotoxic': 746,\n",
       " 'relaxation': 747,\n",
       " 'ago': 748,\n",
       " 'susceptible': 749,\n",
       " 'nos-1': 750,\n",
       " 'instance': 751,\n",
       " 'vasodilation': 752,\n",
       " 'modulation': 753,\n",
       " 'product': 754,\n",
       " 'pathophysiological': 755,\n",
       " 'contend': 756,\n",
       " 'knowledge': 757,\n",
       " 'types': 758,\n",
       " 'continue': 759,\n",
       " 'potential': 760,\n",
       " 'instead': 761,\n",
       " 'immunoreactivity': 762,\n",
       " 'oxidative': 763,\n",
       " 'commonly': 764,\n",
       " 'mouse': 765,\n",
       " 'marker': 766,\n",
       " 'postsinusoidal': 767,\n",
       " 'metabolite': 768,\n",
       " 'relied': 769,\n",
       " 'precise': 770,\n",
       " 'glutaredoxin': 771,\n",
       " 'numerous': 772,\n",
       " 'stimulate': 773,\n",
       " 'illustrates': 774,\n",
       " 'widely': 775,\n",
       " 'led': 776,\n",
       " 'drawback': 777,\n",
       " 'oxidant-dependent': 778,\n",
       " 'nitric': 779,\n",
       " 'no2': 780,\n",
       " 'strategy': 781,\n",
       " 'understanding': 782,\n",
       " 'primarily': 783,\n",
       " 'changes': 784,\n",
       " 'directly': 785,\n",
       " 'note': 786,\n",
       " '6]': 787,\n",
       " 'oxidation': 788,\n",
       " 'activity': 789,\n",
       " 'interpretations': 790,\n",
       " 'decreasing': 791,\n",
       " 'unfortunately': 792,\n",
       " 'tyrosine': 793,\n",
       " 'relates': 794,\n",
       " 'determined': 795,\n",
       " 'interesting': 796,\n",
       " 'contribution': 797,\n",
       " 'inhibition': 798,\n",
       " 'consequent': 799,\n",
       " 'harmful': 800,\n",
       " 'moreover': 801,\n",
       " 'reversed': 802,\n",
       " 'peroxidases': 803,\n",
       " 'assumed': 804,\n",
       " 'unravel': 805,\n",
       " 'antioxidants': 806,\n",
       " 'questionable': 807,\n",
       " 'acid': 808,\n",
       " 's-nitros': 809,\n",
       " 'regulator': 810,\n",
       " 'neutrophils': 811,\n",
       " 'nitration': 812,\n",
       " 'around': 813,\n",
       " 'removal': 814,\n",
       " 'influenza': 815,\n",
       " 'subnormal': 816,\n",
       " 'addition': 817,\n",
       " 'nature': 818,\n",
       " 'radical-induced': 819,\n",
       " 'enzymic': 820,\n",
       " 'venules': 821,\n",
       " 'redox': 822,\n",
       " 'phosphorylation': 823,\n",
       " 'realized': 824,\n",
       " 'pulmonary': 825,\n",
       " 'could': 826,\n",
       " 'mechanism': 827,\n",
       " 'adequately': 828,\n",
       " 'exuberant': 829,\n",
       " 'conclusions': 830,\n",
       " 'collectively': 831,\n",
       " 'pro-inflammatory': 832,\n",
       " 'chronic': 833,\n",
       " 'hocl': 834,\n",
       " 'paraquat': 835,\n",
       " 'tissue': 836,\n",
       " 'airways': 837,\n",
       " 'chemistry': 838,\n",
       " 'analogous': 839,\n",
       " 'animals': 840,\n",
       " 'provide': 841,\n",
       " 'seem': 842,\n",
       " 'neurotransmission': 843,\n",
       " 'o2•-': 844,\n",
       " 'cyclo-oxygenase': 845,\n",
       " '13]': 846,\n",
       " 'epo': 847,\n",
       " 'conductance': 848,\n",
       " 'focus': 849,\n",
       " 'no•': 850,\n",
       " '[25]': 851,\n",
       " 'animal': 852,\n",
       " 'react': 853,\n",
       " 'measured': 854,\n",
       " 's-nitrosothiol': 855,\n",
       " '3-nitrotyrosine': 856,\n",
       " 'nitrogen': 857,\n",
       " 'expressed': 858,\n",
       " 'surveillance': 859,\n",
       " 'expiratory': 860,\n",
       " 'specifically': 861,\n",
       " 'complementary': 862,\n",
       " 'suspects': 863,\n",
       " 'reliable': 864,\n",
       " 'certain': 865,\n",
       " 'exchange': 866,\n",
       " 'nevertheless': 867,\n",
       " 'worsen': 868,\n",
       " 'use': 869,\n",
       " 'molecules': 870,\n",
       " 'better': 871,\n",
       " 'isoforms': 872,\n",
       " 'p21ras': 873,\n",
       " '[16': 874,\n",
       " 'prime': 875,\n",
       " 'gene': 876,\n",
       " 'effects': 877,\n",
       " '33': 878,\n",
       " 'might': 879,\n",
       " 'additional': 880,\n",
       " 'thereby': 881,\n",
       " 'display': 882,\n",
       " 'importantly': 883,\n",
       " 'proliferation': 884,\n",
       " 'administration': 885,\n",
       " 'thus': 886,\n",
       " 'correlated': 887,\n",
       " 'immune': 888,\n",
       " 'feature': 889,\n",
       " 'co2': 890,\n",
       " 'no•-mediated': 891,\n",
       " '26': 892,\n",
       " 'production': 893,\n",
       " 'nos-3': 894,\n",
       " 'cytoskeletal': 895,\n",
       " 'merits': 896,\n",
       " 'alterations': 897,\n",
       " 'insights': 898,\n",
       " 'chemical': 899,\n",
       " 'important': 900,\n",
       " 'metals': 901,\n",
       " 'reduce': 902,\n",
       " 'termed': 903,\n",
       " 'forming': 904,\n",
       " 'enzymes': 905,\n",
       " 'reason': 906,\n",
       " 'reflect': 907,\n",
       " 'distinct': 908,\n",
       " 'phosphatases': 909,\n",
       " '[45]': 910,\n",
       " 'conditions': 911,\n",
       " 'research': 912,\n",
       " 'gas': 913,\n",
       " 'modulate': 914,\n",
       " 'onoo-but': 915,\n",
       " 'necessarily': 916,\n",
       " 'influence': 917,\n",
       " 'endotoxemia': 918,\n",
       " 'suggested': 919,\n",
       " 'differentiation': 920,\n",
       " 'reaction': 921,\n",
       " 'ions': 922,\n",
       " 'impossible': 923,\n",
       " 'unclear': 924,\n",
       " 'clarify': 925,\n",
       " 'immunohistochemical': 926,\n",
       " 'specificity': 927,\n",
       " '[7': 928,\n",
       " 'distinctly': 929,\n",
       " 'deceptively': 930,\n",
       " 'artifacts': 931,\n",
       " '15]': 932,\n",
       " 'indeed': 933,\n",
       " 'free': 934,\n",
       " 'modifications': 935,\n",
       " '[33]': 936,\n",
       " 'characterized': 937,\n",
       " '18': 938,\n",
       " 'interactions': 939,\n",
       " 'o2•-in': 940,\n",
       " 'depend': 941,\n",
       " 'diseases': 942,\n",
       " 'host': 943,\n",
       " 'hypochlorous': 944,\n",
       " 'provocation': 945,\n",
       " 'detect': 946,\n",
       " 'applicable': 947,\n",
       " '36]': 948,\n",
       " 'reverse': 949,\n",
       " 'especially': 950,\n",
       " 'clear': 951,\n",
       " 'preparation': 952,\n",
       " 'paradox': 953,\n",
       " 'linked': 954,\n",
       " 'view': 955,\n",
       " 'presumably': 956,\n",
       " 'hence': 957,\n",
       " 'convenient': 958,\n",
       " 'documented': 959,\n",
       " 'discussed': 960,\n",
       " 'mpo-derived': 961,\n",
       " 'eosinophil': 962,\n",
       " 'suggest': 963,\n",
       " 'local': 964,\n",
       " 'pharmacokinetic': 965,\n",
       " '[16]': 966,\n",
       " 'substantial': 967,\n",
       " 'overwhelming': 968,\n",
       " 'metabolic': 969,\n",
       " 'originating': 970,\n",
       " 'commentary': 971,\n",
       " 'either': 972,\n",
       " 'apoptotic': 973,\n",
       " 'functional': 974,\n",
       " 'aromatic': 975,\n",
       " 'provided': 976,\n",
       " 'yl': 977,\n",
       " 'suitable': 978,\n",
       " 'surface': 979,\n",
       " 'rather': 980,\n",
       " '[24]': 981,\n",
       " 'accepted': 982,\n",
       " 'point': 983,\n",
       " 'mpo': 984,\n",
       " 'iron': 985,\n",
       " '23]': 986,\n",
       " 'rat': 987,\n",
       " 'transmembrane': 988,\n",
       " 'eosinophils': 989,\n",
       " 'need': 990,\n",
       " 'rns': 991,\n",
       " 'limitations': 992,\n",
       " '42]': 993,\n",
       " '[4': 994,\n",
       " 'despite': 995,\n",
       " 'low': 996,\n",
       " 'taking': 997,\n",
       " 'inconsistencies': 998,\n",
       " '[5': 999,\n",
       " ...}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tf_dict(docid):\n",
    "    tf_dict = {}\n",
    "    paths = filemap[docid]\n",
    "    alltext = \"\"\n",
    "    for path in paths.split(\"; \"):\n",
    "        if(path == \"nan\"):\n",
    "            continue\n",
    "        datapath = os.path.join(data_dir,path)\n",
    "        with open(datapath) as data:\n",
    "            data_dict = json.load(data)\n",
    "            for section in data_dict['body_text']:\n",
    "                section_text = section['text']\n",
    "                alltext+=section_text\n",
    "                alltext+=\" \"\n",
    "    tokens = getTokensFromText(alltext)\n",
    "    for token in tokens:\n",
    "        if token not in tf_dict:\n",
    "            tf_dict[token]=1\n",
    "        else:\n",
    "            tf_dict[token]+=1\n",
    "    return tf_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = len(filemap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_idf(term):\n",
    "    return math.log(1+N/idf_dict[term],2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getDocVector(docid):\n",
    "    tf_dict = get_tf_dict(docid)\n",
    "    vec = np.zeros(len(vocab))\n",
    "    for word,tf in tf_dict.items():\n",
    "        if word in vocab:\n",
    "            pos = vocab[word]\n",
    "            normtf = 1 + math.log(tf,2)\n",
    "            normidf = get_idf(word)\n",
    "            tfidf = normtf * normidf\n",
    "            vec[pos] = tfidf\n",
    "    return vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#getDocVector('ug7v899j')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getQueryVector(query):\n",
    "    tokens = getTokensFromText(query)\n",
    "    query_tf_dict = {}\n",
    "    for token in tokens:\n",
    "        if token not in query_tf_dict:\n",
    "            query_tf_dict[token]=1\n",
    "        else:\n",
    "            query_tf_dict[token]+=1\n",
    "    vec = np.zeros(len(vocab))\n",
    "    for word,tf in query_tf_dict.items():\n",
    "        if word in vocab.keys():\n",
    "            pos = vocab[word]\n",
    "            normtf = 1 + math.log(tf,2)\n",
    "            normidf = get_idf(word)\n",
    "            tfidf = normtf * normidf\n",
    "            vec[pos] = tfidf\n",
    "    return vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.299325612080061\n",
      "7.153821821804281\n"
     ]
    }
   ],
   "source": [
    "temp = getQueryVector(\"this is a sample query\")\n",
    "for e in temp:\n",
    "    if e >0:\n",
    "        print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSim(queryVec,docVec):\n",
    "    if np.linalg.norm(queryVec) == 0 or np.linalg.norm(docVec)==0:\n",
    "        return 0.0\n",
    "    return np.dot(queryVec,docVec)/(np.linalg.norm(queryVec) * np.linalg.norm(docVec))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.04362011323686911"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getSim(getQueryVector(\"pneumonia disease breathe\"),getDocVector('ug7v899j'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.07223477347244252"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getSim(getQueryVector(\"inflammatory diseases\"),getDocVector('02tnwd4m'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.03594977410131264"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getSim(getQueryVector(\"disease\"),getDocVector('02tnwd4m'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "topics_file = \"covid19-topics.xml\"\n",
    "top100_file= \"t40-top-100.txt\"\n",
    "relevance_file = \"t40-qrels.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_queries():\n",
    "    with open(os.path.join(data_dir,topics_file), \"r\") as file:\n",
    "        content = file.readlines()\n",
    "        content = \"\".join(content)\n",
    "        bs_content = BeautifulSoup(content, \"lxml\")\n",
    "        queries = bs_content.find_all(\"query\")\n",
    "        all_text = []\n",
    "        for q in queries:\n",
    "            all_text.append(q.text)\n",
    "        return all_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevances():\n",
    "    rel_dict = {}\n",
    "    with open(os.path.join(data_dir,relevance_file),\"r\") as file:\n",
    "        for x in file:\n",
    "            line = x.split(\" \")\n",
    "            qid = int(line[0])\n",
    "            docid = line[2]\n",
    "            rel = int(line[3].replace(\"\\n\",\"\"))\n",
    "            if qid not in rel_dict:\n",
    "                rel_dict[qid] = {docid:rel}\n",
    "            else:\n",
    "                rel_dict[qid][docid] = rel\n",
    "    return rel_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_top100():\n",
    "    top100_dict = {}\n",
    "    with open(os.path.join(data_dir,top100_file),\"r\") as file:\n",
    "        for x in file:\n",
    "            line = x.split(\" \")\n",
    "            if(len(line)>1):\n",
    "                qid = int(line[0])\n",
    "                docid = line[2]\n",
    "                if qid not in top100_dict:\n",
    "                    top100_dict[qid] = [docid]\n",
    "                else:\n",
    "                    top100_dict[qid].append(docid)\n",
    "    return top100_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "top100_dict = get_top100()\n",
    "rel_dict = get_relevances()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test():\n",
    "    docs = top100_dict[1]\n",
    "    for doc in docs:\n",
    "        if doc in rel_dict[1]:\n",
    "            print(rel_dict[1][doc])\n",
    "        else:\n",
    "            print(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getNDCG(qno,docs):\n",
    "    def discount_helper(arr):\n",
    "        score = 0\n",
    "        for i,e in enumerate(arr):\n",
    "            score += e/math.log(i+2,2)\n",
    "        return score\n",
    "    scores = []\n",
    "    for doc in docs:\n",
    "        if doc in rel_dict[qno]:\n",
    "            scores.append(rel_dict[qno][doc])\n",
    "        else:\n",
    "            scores.append(0)\n",
    "    origscore = discount_helper(scores)\n",
    "    sortedscore = discount_helper(sorted(scores,reverse=True))\n",
    "    if sortedscore == 0:\n",
    "        return 0\n",
    "    return origscore/sortedscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7565414085970846"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getNDCG(1,top100_dict[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAvgNDCG():\n",
    "    total = 0\n",
    "    all_scores = []\n",
    "    for i in range(len(top100_dict)):\n",
    "        qno = i+1\n",
    "        qndcg = getNDCG(qno,top100_dict[qno])\n",
    "        all_scores.append(qndcg)\n",
    "        total +=qndcg\n",
    "    return all_scores,total/len(top100_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([0.7565414085970846,\n",
       "  0.5490099597832235,\n",
       "  0.6470841218258803,\n",
       "  0.2591553717297171,\n",
       "  0.5123642355725464,\n",
       "  0.9427036463794559,\n",
       "  0.6847681924597284,\n",
       "  0.49131943659780286,\n",
       "  0,\n",
       "  0.7747093032722049,\n",
       "  0.4086589590455306,\n",
       "  0.6987442782261867,\n",
       "  0.5386631163747705,\n",
       "  0.8480193399948635,\n",
       "  0.6340964575403888,\n",
       "  0.5998317777936504,\n",
       "  0.9103193495961885,\n",
       "  0.9150619381911983,\n",
       "  0.637546728944844,\n",
       "  0.8385409422607322,\n",
       "  0.8639828885427547,\n",
       "  0.8110485560085039,\n",
       "  0.9151046291484798,\n",
       "  0.9751967515222458,\n",
       "  0.852393972418686,\n",
       "  0.8216113839948944,\n",
       "  0.9459717465856947,\n",
       "  0.9769764434729133,\n",
       "  0.8114709918590154,\n",
       "  0.9518500224219281,\n",
       "  0.31163274949398934,\n",
       "  0.34651777214626495,\n",
       "  0.8850340633413392,\n",
       "  0.3310085419934364,\n",
       "  0.7669690018713811,\n",
       "  0.9861473153436482,\n",
       "  0.8387767235757052,\n",
       "  0.9633026448653588,\n",
       "  0.9676811399300267,\n",
       "  0.8986661207975979],\n",
       " 0.7217120505879965)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getAvgNDCG()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rocchio_updateQueries(alpha,beta,gamma):\n",
    "    updated_queries = []\n",
    "    for i,q_orig in tqdm(enumerate(get_queries())):\n",
    "        q = q_orig.lower()\n",
    "        qv0 = getQueryVector(q)\n",
    "        qno = i+1\n",
    "        top100docs = top100_dict[qno]\n",
    "        relvec = np.zeros(len(vocab))\n",
    "        irrelvec = np.zeros(len(vocab))\n",
    "        rel_vecarr = []\n",
    "        irrel_vecarr = []\n",
    "        for doc in top100docs:\n",
    "            if doc in rel_dict[qno]:\n",
    "                rel = rel_dict[qno][doc]\n",
    "            else:\n",
    "                rel = 0\n",
    "            if rel >0:\n",
    "                rel_vecarr.append(getDocVector(doc))\n",
    "            else:\n",
    "                irrel_vecarr.append(getDocVector(doc))\n",
    "        for v in rel_vecarr:\n",
    "            relvec +=v\n",
    "        for v in irrel_vecarr:\n",
    "            irrelvec +=v\n",
    "        if len(rel_vecarr)>0:\n",
    "            relvec /= len(rel_vecarr)\n",
    "        if len(irrel_vecarr)>0:\n",
    "            irrelvec /= len(irrel_vecarr)\n",
    "        qnew = np.multiply(qv0,alpha) +np.multiply(relvec,beta) - np.multiply(irrelvec,gamma)\n",
    "        qnew = np.where(qnew<0, 0, qnew)\n",
    "        updated_queries.append(qnew)\n",
    "    return updated_queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1e7a449959b047fd879c818a7e25dc75",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "updated_queries = rocchio_updateQueries(1.0,0.75,0.15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rocchio_getNDCG(updated_queries):\n",
    "    total = 0\n",
    "    allscores = []\n",
    "    for i,q in tqdm(enumerate(updated_queries)):\n",
    "        qno = i+1\n",
    "        doc_score_dict = {}\n",
    "        top100docs = top100_dict[qno]\n",
    "        for doc in top100docs:\n",
    "            score = getSim(q,getDocVector(doc))\n",
    "            doc_score_dict[doc] = score\n",
    "        sorted_docs = dict(sorted(doc_score_dict.items(), key=lambda x: x[1],reverse = True))\n",
    "        sorted_docs = list(sorted_docs.keys())\n",
    "        ndcg = getNDCG(qno,sorted_docs)\n",
    "        allscores.append(ndcg)\n",
    "        total +=ndcg\n",
    "    return allscores,total/len(updated_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "60fe0f0c5398457d80c08b89869c3e86",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.8635512790394195"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rocchio_getNDCG(updated_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gridSearch():\n",
    "    res = {'alpha':[],'beta':[],'gamma':[],'qid':[],'nDCG':[]}\n",
    "    avgres = {'alpha':[],'beta':[],'gamma':[],'nDCG':[]}\n",
    "    df = pd.DataFrame(res)\n",
    "    avgdf = pd.DataFrame(avgres)\n",
    "    for alpha in [1.0]:\n",
    "        for beta in [0.75]:\n",
    "            for gamma in [0.15]:\n",
    "                updated_queries = rocchio_updateQueries(alpha,beta,gamma)\n",
    "                allscores,ndcg = rocchio_getNDCG(updated_queries)\n",
    "                alpha_arr = [alpha] * len(updated_queries)\n",
    "                beta_arr = [beta] * len(updated_queries)\n",
    "                gamma_arr = [gamma] * len(updated_queries)\n",
    "                q_arr = np.arange(1,len(updated_queries)+1,1)\n",
    "                df2 = pd.DataFrame({'alpha':alpha_arr,'beta':beta_arr,'gamma':gamma_arr,'qid':q_arr,'nDCG':allscores})\n",
    "                df = df.append(df2,ignore_index=True)\n",
    "                avgdf = avgdf.append(pd.DataFrame({'alpha':[alpha],'beta':[beta],'gamma':[gamma],'nDCG':[ndcg]}))\n",
    "    df.to_csv('query_wise.csv',index=False)\n",
    "    avgdf.to_csv('avg.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "66b5fd60cda1416d9243ed2c1cc3511b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d4d4900554174f97a8d93833162fd25b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "gridSearch()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getRandomDocs(num):\n",
    "    elems = random.sample(list(filemap),num)\n",
    "    return elems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rocchio_updateQueriesTest(alpha,beta,gamma):\n",
    "    updated_queries = []\n",
    "    irrelvec = np.zeros(len(vocab))\n",
    "    randomdocs = getRandomDocs(100)\n",
    "    irrel_vecarr = []\n",
    "    for doc in randomdocs:\n",
    "        irrel_vecarr.append(getDocVector(doc))\n",
    "    for v in irrel_vecarr:\n",
    "        irrelvec +=v\n",
    "    if len(irrel_vecarr)>0:\n",
    "        irrelvec /= len(irrel_vecarr)\n",
    "    for i,q_orig in tqdm(enumerate(get_queries())):\n",
    "        q = q_orig.lower()\n",
    "        qv0 = getQueryVector(q)\n",
    "        qno = i+1\n",
    "        top100docs = top100_dict[qno]\n",
    "        relvec = np.zeros(len(vocab))\n",
    "        rel_vecarr = []\n",
    "        for doc in top100docs:\n",
    "            rel_vecarr.append(getDocVector(doc))\n",
    "        for v in rel_vecarr:\n",
    "            relvec +=v\n",
    "        if len(rel_vecarr)>0:\n",
    "            relvec /= len(rel_vecarr)\n",
    "        qnew = np.multiply(qv0,alpha) +np.multiply(relvec,beta) - np.multiply(irrelvec,gamma)\n",
    "        qnew = np.where(qnew<0, 0, qnew)\n",
    "        updated_queries.append(qnew)\n",
    "    return updated_queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "28404ae282f74fd09f052fa8814a4c33",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "updated_queries = rocchio_updateQueriesTest(1.0,0.75,0.15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rocchio_getNDCG(updated_queries):\n",
    "    total = 0\n",
    "    allscores = []\n",
    "    for i,q in tqdm(enumerate(updated_queries)):\n",
    "        qno = i+1\n",
    "        doc_score_dict = {}\n",
    "        top100docs = top100_dict[qno]\n",
    "        for doc in top100docs:\n",
    "            score = getSim(q,getDocVector(doc))\n",
    "            doc_score_dict[doc] = score\n",
    "        sorted_docs = dict(sorted(doc_score_dict.items(), key=lambda x: x[1],reverse = True))\n",
    "        sorted_docs = list(sorted_docs.keys())\n",
    "        ndcg = getNDCG(qno,sorted_docs)\n",
    "        allscores.append(ndcg)\n",
    "        total +=ndcg\n",
    "    return allscores,total/len(updated_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rocchio_getNDCG' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-45-9afb52eb330c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mrocchio_getNDCG\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mupdated_queries\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'rocchio_getNDCG' is not defined"
     ]
    }
   ],
   "source": [
    "rocchio_getNDCG(updated_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getFinal(updated_queries):\n",
    "    res_file = open(\"outputt.txt\",\"w\")\n",
    "    for i,q in tqdm(enumerate(updated_queries)):\n",
    "        qno = i+1\n",
    "        doc_score_dict = {}\n",
    "        top100docs = top100_dict[qno]\n",
    "        for doc in top100docs:\n",
    "            score = getSim(q,getDocVector(doc))\n",
    "            doc_score_dict[doc] = score\n",
    "        sorted_docs = dict(sorted(doc_score_dict.items(), key=lambda x: x[1],reverse = True))\n",
    "        for r,doc in enumerate(sorted_docs.keys()):\n",
    "            rank = r+1\n",
    "            score = sorted_docs[doc]\n",
    "            l = str(qno) + \" Q0 \"+ doc +\" \" + str(rank) + \" \" + str(score) + \" \" + \"runid1\\n\"\n",
    "            res_file.write(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a3be76fcd57d404d980c9d9099900da0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=1.0, bar_style='info', max=1.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "getFinal(updated_queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
